source/main.cpp

```diff
@@ -0,0 +1,248 @@
+        }
+
+        ret = receive_test_data();
+
+        if (ret < 0) {
+            printf("%sSocket.recv() fails, code: %d\n", SOCKET_TYPE, ret);
+            return false;
+        } else {
+            printf("Received from echo server %d Bytes\n", ret);
+        }
+
+        _socket.close();
+
+        if (ret != NSAPI_ERROR_OK) {
+            printf("%sSocket.close() fails, code: %d\n", SOCKET_TYPE, ret);
+            return false;
+        }
+
+        return true;
+    }
+
+    /** Connects to the Cellular Network */
+    bool connect_cellular()
+    {
+        printf("Establishing connection\n");
+
+        /* check if we're already connected */
+        if (_net.get_connection_status() == NSAPI_STATUS_GLOBAL_UP) {
+            return true;
+        }
+
+        nsapi_error_t ret;
+
+        for (uint8_t retry = 0; retry <= RETRY_COUNT; retry++) {
+            ret = _net.connect();
+
+            if (ret == NSAPI_ERROR_OK) {
+                printf("Connection Established.\n");
+                return true;
+            } else if (ret == NSAPI_ERROR_AUTH_FAILURE) {
+                printf("Authentication Failure.\n");
+                return false;
+            } else {
+                printf("Couldn't connect: %d, will retry\n", ret);
+            }
+        }
+
+        printf("Fatal connection failure: %d\n", ret);
+
+        return false;
+    }
+
+    /** Connects to the Cellular Network */
+    bool resolve_hostname()
+    {
+#if MBED_CONF_APP_SOCK_TYPE != NONIP
+        nsapi_error_t ret = _net.gethostbyname(ECHO_HOSTNAME, &_socket_address);
+
+        if (ret != NSAPI_ERROR_OK) {
+            printf("Couldn't resolve remote host: %s, code: %d\n", ECHO_HOSTNAME, ret);
+            return false;
+        }
+
+        _socket_address.set_port(MBED_CONF_APP_ECHO_SERVER_PORT);
+#endif
+        return true;
+    }
+
+    bool connect_socket()
+    {
+#if MBED_CONF_APP_SOCK_TYPE == TCP
+        nsapi_error_t ret = _socket.connect(_socket_address);
+        if (ret < 0) {
+            printf("TCPSocket.connect() fails, code: %d\n", ret);
+            return false;
+        } else {
+            printf("TCP: connected with %s server\n", ECHO_HOSTNAME);
+        }
+#endif
+        return true;
+    }
+
+    nsapi_error_t send_test_data()
+    {
+        const char *echo_string = "TEST";
+#if MBED_CONF_APP_SOCK_TYPE == UDP
+        return _socket.sendto(_socket_address, (void*)echo_string, strlen(echo_string));
+#else
+        return _socket.send((void*)echo_string, strlen(echo_string));
+#endif
+    }
+
+    nsapi_error_t receive_test_data()
+    {
+        char receive_buffer[4];
+#if MBED_CONF_APP_SOCK_TYPE == UDP
+        return _socket.recvfrom(&_socket_address, (void*)receive_buffer, sizeof(receive_buffer));
+#else
+        return _socket.recv((void*)receive_buffer, sizeof(receive_buffer));
+#endif
+    }
+
+private:
+    NetworkInterface &_net;
+
+#if MBED_CONF_APP_SOCK_TYPE == TCP
```

Comment:
Seems like a shame that we didn't use templates for the `Socket` API, then we probably could've templated on the packet type instead of doing this.